lib/ykpiv.c

```diff
@@ -202,6 +203,53 @@ ykpiv_rc ykpiv_connect(ykpiv_state *state, const char *wanted) {
   return YKPIV_GENERIC_ERROR;
 }
 
+ykpiv_rc ykpiv_reconnect(ykpiv_state *state) {
+  unsigned long active_protocol;
+  long rc;
+
+  if(state->verbose) {
+    fprintf(stderr, "trying to reconnect to current reader.\n");
+  }
+  rc = SCardReconnect(state->card, SCARD_SHARE_SHARED,
+    SCARD_PROTOCOL_T1, SCARD_RESET_CARD, &active_protocol);
+  if(rc != SCARD_S_SUCCESS) {
+    if(state->verbose) {
+      fprintf(stderr, "SCardReconnect failed, rc=%08lx\n", rc);
+    }
+    return YKPIV_PCSC_ERROR;
+  }
+
+  {
+    APDU apdu;
+    unsigned char data[0xff];
+    unsigned long recv_len = sizeof(data);
+    int sw, tries;
+    ykpiv_rc res;
+
+    memset(apdu.raw, 0, sizeof(apdu));
+    apdu.st.ins = 0xa4;
+    apdu.st.p1 = 0x04;
+    apdu.st.lc = sizeof(aid);
+    memcpy(apdu.st.data, aid, sizeof(aid));
+
+    if((res = send_data(state, &apdu, data, &recv_len, &sw)) != YKPIV_OK) {
+      if(state->verbose) {
+        fprintf(stderr, "Failed communicating with card: '%s'\n", ykpiv_strerror(res));
+      }
+      return YKPIV_PCSC_ERROR;
+    } else if(sw == SW_SUCCESS) {
+      return ykpiv_verify(state, state->pin, &tries);
```

Comment:
this might be problematic if state->pin is not set yet, probably needs a check for that.